x-pack/plugins/viz_editor/pie_chart_plugin/pie_chart_plugin.tsx

```diff
@@ -0,0 +1,158 @@
+ */
+
+// @ts-ignore
+import { EuiSuperSelect } from '@elastic/eui';
+import clone from 'lodash-es/clone';
+import setWith from 'lodash-es/setWith';
+import React from 'react';
+import { columnSummary } from '../common/components/config_panel';
+import { IndexPatternPanel } from '../common/components/index_pattern_panel';
+import { Axis, selectColumn, ViewModel } from '../common/lib';
+import { EditorPlugin, PanelComponentProps } from '../public/editor_plugin_registry';
+
+interface PieChartPrivateState {
+  sliceAxis: Axis;
+  angleAxis: Axis;
+}
+
+type PieChartViewModel = ViewModel<'pieChart', PieChartPrivateState>;
+
+function dataPanel({ viewModel, onChangeViewModel }: PanelComponentProps<PieChartViewModel>) {
+  return (
+    <IndexPatternPanel
+      indexPatterns={viewModel.indexPatterns}
+      onChangeIndexPatterns={indexPatterns => {
+        onChangeViewModel({ ...viewModel, indexPatterns });
+      }}
+    />
+  );
+}
+
+function configPanel({ viewModel, onChangeViewModel }: PanelComponentProps<PieChartViewModel>) {
+  const {
+    private: {
+      pieChart: { sliceAxis, angleAxis },
+    },
+  } = viewModel;
+  return (
+    <>
+      <div className="configPanel-axis">
+        <span className="configPanel-axis-title">Slice pie by</span>
+        {sliceAxis.columns.map(col => (
+          <span>{columnSummary(selectColumn(col, viewModel))}</span>
+        ))}
+      </div>
+      <div className="configPanel-axis">
+        <span className="configPanel-axis-title">Size slices by</span>
+        {angleAxis.columns.map(col => (
+          <span>{columnSummary(selectColumn(col, viewModel))}</span>
+        ))}
+      </div>
+    </>
+  );
+}
+
+function toExpression(viewState: PieChartViewModel) {
+  const legacyConfig = {
+    type: 'pie',
+    addTooltip: true,
+    addLegend: true,
+    legendPosition: 'right',
+    isDonut: true,
+    labels: {
+      show: false,
+      values: true,
+      last_level: true,
+      truncate: 100,
+    },
+    dimensions: {
+      metric: {
+        accessor: 1,
+        format: {
+          id: 'number',
+        },
+        params: {},
+        aggType: 'count',
+      },
+      buckets: [
+        {
+          accessor: 0,
+          format: {
+            id: 'terms',
+            params: {
+              id: 'string',
+              otherBucketLabel: 'Other',
+              missingBucketLabel: 'Missing',
+            },
+          },
+          params: {},
+          aggType: 'terms',
+        },
+      ],
+    },
+  };
+  // TODO prob. do this on an AST object and stringify afterwards
+  // TODO actually use the stuff from the viewState
+  return `sample_data | pie_chart | kibana_pie visConfig='${JSON.stringify(legacyConfig)}'`;
+}
+
+function prefillPrivateState(viewModel: ViewModel<string, unknown>) {
+  if (viewModel.private.pieChart) {
+    return viewModel;
+  }
+
+  // TODO we maybe need a more stable way to get these
+  const xAxisRef = 'q1_0';
+  const yAxisRef = 'q1_1';
```

Comment:
I imagine the simpler way of doing this is a helper function such as:

```
function getRefByType(type: string) {
  return state.queries.filter(({ queryType }) => queryType === type);
}
```

Reply:
I agree, most of the stuff which is done in the prefill and suggestion functions should be handled by helper functions out of the common folder - as long as they are pure functions this should scale quite well.

But we don't have a queryType on the queries yet, right?